src/plugins/event_annotation/common/fetch_event_annotations/request_event_annotations.ts

```diff
@@ -0,0 +1,352 @@
+    return {
+      esaggsParams: { dataView, aggConfigs, timeFields: [group.timeField] },
+      fieldsColIdMap:
+        group.allFields?.reduce<Record<string, string>>(
+          (acc, fieldName, i) => ({
+            ...acc,
+            [fieldName]: `col-${i + 4}-${i + 5}`,
+          }),
+          {}
+        ) || {},
+    };
+  });
+};
+
+function regroupForRequestOptimization(
+  { groups }: FetchEventAnnotationsArgs,
+  input: ExpressionValueSearchContext | null
+) {
+  const outputGroups = groups
+    .map((g) => {
+      return g.annotations.reduce<Record<string, ManualGroup | QueryGroup>>((acc, current) => {
+        if (current.isHidden) {
+          return acc;
+        }
+
+        if (isManualAnnotation(current)) {
+          if (!isInRange(current, input?.timeRange)) {
+            return acc;
+          }
+          const manualSubgroup = acc.manual as ManualGroup | undefined;
+          return {
+            ...acc,
+            manual: {
+              type: 'manual',
+              annotations: [...(manualSubgroup ? manualSubgroup.annotations : []), current],
+            },
+          };
+        } else {
+          const key = `${g.dataView.value.id}-${current.timeField}`;
+          const subGroup = acc[key] as QueryGroup;
+          if (subGroup) {
+            let allFields = [...(subGroup.allFields || []), ...(current.extraFields || [])];
+            if (current.textField) {
+              allFields = [...allFields, current.textField];
+            }
+            return {
+              ...acc,
+              [key]: {
+                ...subGroup,
+                allFields: [...new Set(allFields)],
+                annotations: [...subGroup.annotations, current],
+              },
+            };
+          }
+
+          let allFields = current.extraFields || [];
+          if (current.textField) {
+            allFields = [...allFields, current.textField];
+          }
+
+          return {
+            ...acc,
+            [key]: {
+              type: 'query',
+              dataView: g.dataView,
+              timeField: current.timeField,
+              allFields,
+              annotations: [current],
+            },
+          };
+        }
+      }, {});
+    })
+    .reduce((acc, currentGroup) => {
+      Object.keys(currentGroup).forEach((key) => {
+        if (acc[key]) {
+          const currentSubGroup = currentGroup[key];
+          const requestGroup = acc[key];
+
+          if (isManualSubGroup(currentSubGroup) || isManualSubGroup(requestGroup)) {
+            acc[key] = {
+              ...requestGroup,
+              annotations: [...requestGroup.annotations, ...currentSubGroup.annotations],
+            } as ManualGroup;
+          } else {
+            acc[key] = {
+              ...requestGroup,
+              annotations: [...requestGroup.annotations, ...currentSubGroup.annotations],
+              allFields: [...(requestGroup.allFields || []), ...(currentSubGroup.allFields || [])],
```

Comment:
Here the code could dedup once for all with a:
```suggestion
              allFields: [
                ...new Set([
                  ...(requestGroup.allFields || []),
                  ...(currentSubGroup.allFields || []),
                ]),
              ],
```